exporter/exporterhelper/wal.go

```diff
@@ -0,0 +1,305 @@
+		wcs.writeIndex = 0
+	} else {
+		val, conversionErr := bytesToUint64(writeIndexBytes)
+		if conversionErr != nil {
+			// TODO: consider failing here?
+			wcs.logger.Warn("write index corrupted, starting with a new one", zap.String(zapWalIDKey, wcs.id))
+			wcs.writeIndex = 0
+		} else {
+			wcs.writeIndex = val
+		}
+	}
+}
+
+// put marshals the request and puts it into the WAL
+func (wcs *walContiguousStorage) put(ctx context.Context, req request) error {
+	buf, err := req.marshall()
+	if err != nil {
+		return err
+	}
+
+	wcs.mu.Lock()
+	defer wcs.mu.Unlock()
+
+	itemKey := wcs.buildItemKey(wcs.writeIndex)
+	err = wcs.client.Set(ctx, itemKey, buf)
+	if err != nil {
+		return err
+	}
+
+	wcs.writeIndex++
+	writeIndexBytes, err := uint64ToBytes(wcs.writeIndex)
+	if err != nil {
+		wcs.logger.Warn("failed converting write index uint64 to bytes", zap.Error(err))
+	}
+
+	return wcs.client.Set(ctx, wcs.writeIndexKey, writeIndexBytes)
+}
+
+// get returns the next request from the queue; note that it might be blocking if there are no entries in the WAL
+func (wcs *walContiguousStorage) get(ctx context.Context) (request, error) {
+	wcs.mu.Lock()
+	defer wcs.mu.Unlock()
+
+	if wcs.readIndex >= wcs.writeIndex {
+		return nil, errNotFound
+	}
+
+	itemKey := wcs.buildItemKey(wcs.readIndex)
+	// Increase here, so despite errors it would still progress
+	wcs.readIndex++
+
+	buf, err := wcs.client.Get(ctx, itemKey)
+	if err != nil {
+		// This is the only critical error
+		return nil, err
+	}
+
+	// This needs to happen after unmarshalling as buf contents might change
+	defer wcs.updateItemRead(ctx, itemKey)
+
+	return wcs.unmarshaler(buf)
+}
+
+func (wcs *walContiguousStorage) size() int {
+	return int(wcs.writeIndex - wcs.readIndex)
+}
+
+func (wcs *walContiguousStorage) stop() {
+	wcs.logger.Debug("stopping walContiguousStorage", zap.String(zapWalIDKey, wcs.id))
+}
+
+func (wcs *walContiguousStorage) updateItemRead(ctx context.Context, itemKey string) {
+	err := wcs.client.Delete(ctx, itemKey)
+	if err != nil {
+		wcs.logger.Debug("failed deleting item", zap.String(zapItemKey, itemKey))
+	}
+
+	readIndexBytes, err := uint64ToBytes(wcs.readIndex)
+	if err != nil {
+		wcs.logger.Warn("failed converting read index uint64 to bytes", zap.Error(err))
+	} else {
+		err = wcs.client.Set(ctx, wcs.readIndexKey, readIndexBytes)
+		if err != nil {
+			wcs.logger.Warn("failed storing read index", zap.Error(err))
+		}
+	}
+}
+
+func (wcs *walContiguousStorage) buildItemKey(index uint64) string {
+	return fmt.Sprintf(itemKeyTemplate, wcs.id, index)
+}
+
+func (wcs *walContiguousStorage) buildReadIndexKey() string {
+	return fmt.Sprintf(readIndexTemplate, wcs.id)
+}
+
+func (wcs *walContiguousStorage) buildWriteIndexKey() string {
+	return fmt.Sprintf(writeIndexTemplate, wcs.id)
+}
+
+func uint64ToBytes(val uint64) ([]byte, error) {
+	buf := new(bytes.Buffer)
+	err := binary.Write(buf, binary.LittleEndian, val)
```

Comment:
Escape analysis might optimize this away but you can explicitly put it on stack:

```suggestion
	var buf bytes.Buffer
	err := binary.Write(&buf, binary.LittleEndian, val)
```

Reply:
Oh, sweet!